Gallium starts local protected http server

ernesto-jimenez opened the pull request:
With these changes:
- gallium.App will have a local HTTP server running on a free port.
- the server is protected with a random token generated on startup.
- it exposes http.DefaultServerMux

Re: #16

Next step is to allow gallium to send events to the webpage through that server.

ernesto-jimenez commented:
@alexflint I've included a `webapp` example that serves one of the To Do apps from http://todomvc.com from the embedded server having the assets embedded in the binary with `go-bindata-assetfs`

alexflint commented on app.go:
```diff
@@ -60,7 +61,8 @@ func Loop(args []string, onReady func(*App)) error {
 	app := App{
-		ready: make(chan struct{}),
+		ready:  make(chan struct{}),
+		server: newServer(http.DefaultServeMux),
```
Hmm won't using `DefaultServeMux` here make it impossible to use third party http handlers, like `gorilla.Mux` etc? My suggestion would be:
- export `newServer` as `gallium.NewServer`
- require users to instantiate their own server if they want one
- require that users pass in their local server url explicitly as in `OpenWindow(myserver.BaseURL(), opts)`

That will also mean that users will have the flexibility to avoid opening any network ports at all if they don't want a server, or to have multiple servers if they want that (e.g. one per window). What do you think?

ernesto-jimenez replied:
> Hmm won't using DefaultServeMux here make it impossible to use third party http handlers, like gorilla.Mux etc?

Those packages usually implement the `http.Handler` interface, which means they can register with `DefaultServeMux`

> My suggestion would be:
> 
> * export newServer as gallium.NewServer
> * require users to instantiate their own server if they want one
> * require that users pass in their local server url explicitly as in OpenWindow(myserver.BaseURL(), opts)
> 
> That will also mean that users will have the flexibility to avoid opening any network ports at all if they don't want a server, or to have multiple servers if they want that (e.g. one per window). What do you think?

If we go with the approach in PR #19, we could do the following:

* By default, start the HTTP server using `http.DefaultServeMux`
* Allow apps to implement the following interface `type HandlerHTTP interface { HandlerHTTP() http.Handler }`  to set a custom server to use or disable the local server entirely.
* `OpenWindow`, when called with a relative URL, would load the path in the local server with the authentication parameter.

## Using DefaultServeMux by default


```go
package webapp

func init() {
    HandleFunc("/", handler)
}

type webapp struct {}

func New() *webapp {
    return &webapp{}
}

func (*webapp) Start(app *gallium.App) {
    app.OpenWindow("/", opts)
}
```

## Using custom server

```go
package webapp

type webapp struct {
    server *ServeMux
}

func New() *webapp {
    s := NewServeMux()
    s.HandleFunc("/", handler)
    return &webapp{server: s}
}

func (*webapp) Start(app *gallium.App) {
    app.OpenWindow("/", opts)
}

func (w *webapp) HandlerHTTP() http.Handler {
    return w.server
}
```

## Disabling the HTTP Server

```go
package webapp

type webapp struct {}

func New() *webapp {
    return &webapp{}
}

func (*webapp) Start(app *gallium.App) {
    app.OpenWindow("http://example.com", opts)
}

func (w *webapp) HandlerHTTP() http.Handler {
    return nil
}
```

alexflint replied:
I see where you're coming from, but I still think that the http server should be decoupled from window management. The reason, I guess, is that it doesn't seem like the benefits are large enough to make it worth having them coupled. There definitely are benefits to having them coupled, but I just don't think those benefits are large enough. It should be fairly painless for the user to construct full URLs, and creating your own serve mux is a fairly standard thing that most go developers who have build web apps do often.

joeblew99 replied:
can this be merged :) Its just an example... Can be changed / deleted later

ernesto-jimenez commented on server.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+func token() string {
+	p := make([]byte, 128)
```
👍 will change that

ernesto-jimenez commented on server.go:
```diff
@@ -0,0 +1,106 @@
+	r.URL.RawQuery = params.Encode()
+	r.RequestURI = r.URL.RequestURI()
+
+	// Stripe cookie token
```
Me and my typos... Will change that :)

ernesto-jimenez commented on server.go:
```diff
@@ -0,0 +1,106 @@
+
+	// Stripe cookie token
+	cookies := r.Cookies()[:]
+	r.Header.Del("Cookie")
```
The token should only be a concern of gallium. Never of the app. If there is a problem with the cookie the requests will never get to the app.

I would rather keep this within gallium and avoid leaking it to the app. That way we avoid developers relying on that cookie, which would allow us to change if we need to.

ernesto-jimenez commented:
@alexflint I addressed your comments.

What do you think about changing this so `OpenWindow` accepts relative URLs that would load on the local server?

Allowing:

```go
app.OpenWindow("/settings", opts)
```

Also, once we go forward with PR #19 I can change the code to allow using custom HTTP handlers and disabling the HTTP server, [as described in the previous comment](https://github.com/alexflint/gallium/pull/20#discussion_r82512450)

joeblew99 commented:
If you look at the golang Goa project, it shows how http servers can be designed and then be reused by other http mixers.
There are a few reference to this in their issues list.

It works really well I have found

joeblew99 commented:
https://github.com/goadesign
